libraries/AP_Mount/AP_Mount_Gremsy.cpp

```diff
@@ -0,0 +1,318 @@
+            }
+            _state._roi_target = AP::ahrs().get_home();
+            _state._roi_target_set = true;
+            if (calc_angle_to_roi_target(_angle_ef_target_rad, true, true, false)) {
+                send_gimbal_device_set_attitude(_angle_ef_target_rad.x, _angle_ef_target_rad.y, _angle_ef_target_rad.z, true);
+            }
+            break;
+
+        case MAV_MOUNT_MODE_SYSID_TARGET:
+            if (calc_angle_to_sysid_target(_angle_ef_target_rad, true, true, false)) {
+                send_gimbal_device_set_attitude(_angle_ef_target_rad.x, _angle_ef_target_rad.y, _angle_ef_target_rad.z, true);
+            }
+            break;
+
+        default:
+            // unknown mode so do nothing
+            break;
+    }
+}
+
+// send_mount_status - called to allow mounts to send their status to GCS using the MOUNT_STATUS message
+void AP_Mount_Gremsy::send_mount_status(mavlink_channel_t chan)
+{
+    // check we have space for the message
+    if (!HAVE_PAYLOAD_SPACE(chan, GIMBAL_DEVICE_ATTITUDE_STATUS)) {
+        return;
+    }
+
+    // check we have received an updated message
+    if (_gimbal_device_attitude_status.time_boot_ms == _sent_gimbal_device_attitude_status_ms) {
+        return;
+    }
+    _sent_gimbal_device_attitude_status_ms = _gimbal_device_attitude_status.time_boot_ms;
+
+    // forward on message to GCS
+    mavlink_msg_gimbal_device_attitude_status_send(chan,
+                                                   0,    // target system
+                                                   0,    // target component
+                                                   AP_HAL::millis(),    // autopilot system time
+                                                   _gimbal_device_attitude_status.flags,
+                                                   _gimbal_device_attitude_status.q,
+                                                   _gimbal_device_attitude_status.angular_velocity_x,
+                                                   _gimbal_device_attitude_status.angular_velocity_y,
+                                                   _gimbal_device_attitude_status.angular_velocity_z,
+                                                   _gimbal_device_attitude_status.failure_flags);
+}
+
+// search for gimbal in GCS_MAVLink routing table
+void AP_Mount_Gremsy::find_gimbal()
+{
+    // do not look for gimbal for first 10 seconds so user may see banner
+    if (AP_HAL::millis() < 10000) {
+        return;
+    }
+
+    // return if search time has has passed
+    if (AP_HAL::millis() > AP_MOUNT_GREMSY_SEARCH_MS) {
+        return;
+    }
+
+    // search for a mavlink enabled gimbal
+    if (!_found_gimbal) {
+        mavlink_channel_t chan;
+        uint8_t sysid, compid;
+        if (GCS_MAVLINK::find_by_mavtype(MAV_TYPE_GIMBAL, sysid, compid, chan)) {
+            if (((_instance == 0) && (compid == MAV_COMP_ID_GIMBAL)) ||
+                ((_instance == 1) && (compid == MAV_COMP_ID_GIMBAL2))) {
```

Comment:
I'm a bit puzzled, are you sure this is working correctly? Wouldn't find_by_mavtype() return the same compid for both instances, that is, wouldn't the result depend on a race of which gimbal is noted first?
I found I needed to do this differently, using an extended version of find_by_mavtype() which also accounts for the desired compid, see https://github.com/olliw42/BetaPilot/blob/52d6883f68c01aa0987b56618c1a29f7fa13f01c/libraries/AP_Mount/BP_Mount_STorM32_MAVLink.cpp#L813-L817

Reply:
Good find, thanks.  What we have works for a single gimbal but you're right that it's not going to work well if there are two gimbals on the same serial port.  For now I'm putting off the two-gimbal support for a future PR, I'm sure lots of issues will be uncovered when we test that.